substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/BundleSupport.java

```diff
@@ -182,6 +284,190 @@ static BundleSupport create(NativeImage nativeImage, String bundleArg, NativeIma
+            return null;
+        } catch (NoSuchAlgorithmException e) {
+            throw GraalError.shouldNotReachHere(e); // ExcludeFromJacocoGeneratedReport
+        }
+    }
+
+    private int createContainer() {
+        ProcessBuilder pbCheckForImage = new ProcessBuilder(containerTool, "images", "-q", containerImage + ":latest");
+        ProcessBuilder pb = new ProcessBuilder(containerTool, "build", "-f", dockerfile.toString(), "-t", containerImage, ".");
+        Process p = null;
+        try {
+            p = pbCheckForImage.start();
+            p.waitFor();
+            String imageId = (new BufferedReader(new InputStreamReader(p.getInputStream()))).readLine();
+            if(imageId == null) {
+                pb.inheritIO();
+            } else {
+                nativeImage.showMessage(String.format("%sReusing container image %s.", BUNDLE_INFO_MESSAGE_PREFIX, containerImage));
+            }
+            p = pb.start();
+            int status = p.waitFor();
+            if(imageId != null && status == 0) {
+                Stream<String> result = (new BufferedReader(new InputStreamReader(p.getInputStream()))).lines();
+                p = pbCheckForImage.start();
+                p.waitFor();
+                if(!(new BufferedReader(new InputStreamReader(p.getInputStream()))).readLine().equals(imageId)) {
+                    nativeImage.showMessage(String.format("%sUpdated container image %s.", BUNDLE_INFO_MESSAGE_PREFIX, containerImage));
+                    result.forEach(System.out::println);
+                }
+            }
+            return status;
+        } catch (IOException | InterruptedException e) {
+            throw NativeImage.showError(e.getMessage());
+        } finally {
+            if (p != null) {
+                p.destroy();
+            }
+        }
+    }
+
+    private boolean isToolAvailable(String tool) {
+        String[] paths = SubstrateUtil.split(System.getenv("PATH"), ":");
+        for (String path : paths) {
+            try (Stream<Path> walk = Files.walk(Path.of(path))) {
+                if(walk.anyMatch(p -> p.endsWith(tool))) {
+                    return true;
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+        return false;
+    }
+
+    private String getContainerToolVersion(String tool) {
+        ProcessBuilder pb = new ProcessBuilder(tool, "--version");
+        Process p = null;
+        try {
+            p = pb.start();
+            p.waitFor();
+            String result = (new BufferedReader(new InputStreamReader(p.getInputStream()))).readLine();
+            return result.contains("version") ? result : null;
+        } catch (IOException | InterruptedException e) {
+            throw NativeImage.showError(e.getMessage());
+        } finally {
+            if (p != null) {
+                p.destroy();
+            }
+        }
+    }
+
+    private boolean isRootlessDocker() {
```

Comment:
isRootlessDocker and getContainerToolVersion both reuse a common private method.